src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,501 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+                co_await _as.request_abort();
+                co_await _workers_gate.close();
+                co_await _as.stop();
+                co_return;
+            }
+
+            std::vector<shard_fetch> completed_shard_fetches = std::move(
+              _completed_shard_fetches);
+            _completed_shard_fetches.clear();
+
+            for (auto& sf : completed_shard_fetches) {
+                auto shard = sf.shard;
+                ssx::spawn_with_gate(
+                  _workers_gate,
+                  [this, &octx, shard, sf = std::move(sf)]() mutable {
+                      return log_and_ignore_exceptions(start_shard_fetch_worker(
+                        shard,
+                        octx,
+                        std::move(sf),
+                        // Require that a worker returns more data than before.
+                        // Otherwise it'll return right away with the previous
+                        // result.
+                        _last_result_size[shard] + 1));
+                  });
+            }
+        }
+    }
+
+private:
+    /**
+     * Determines if the should_stop_fetch() condition should be checked again.
+     * The return future is set if;
+     * - octx.deadline has been reached.
+     * - _as has been aborted.
+     * - one of the shard workers has returned results.
+     */
+    ss::future<> progress_conditions(op_context& octx) {
+        ss::promise<> timeout_pr, abort_pr;
+
+        auto abort_sub_opt = _as.subscribe(
+          [&abort_pr]() noexcept { abort_pr.set_value(); });
+
+        if (!abort_sub_opt) {
+            abort_pr.set_value();
+        }
+
+        ss::timer<model::timeout_clock> s{
+          [&timeout_pr] { timeout_pr.set_value(); }};
+
+        if (octx.deadline) {
+            s.arm(octx.deadline.value());
+        }
+
+        co_await ss::when_any(
+          ignore_exceptions(_has_data.wait()),
+          ignore_exceptions(timeout_pr.get_future()),
+          ignore_exceptions(abort_pr.get_future()));
+    }
+
+    /*
+     * `start_shard_fetch_worker` executes on the coordinator shard. It builds
+     * the `shard_local_fetch_context` struct needed to start the
+     * shard_fetch_worker then makes a cross shard call to the shard fetch
+     * worker if needed.
+     *
+     * It then waits until the shard_fetch_worker completes
+     * and then notifies the fetch coordinator.
+     */
+    ss::future<> start_shard_fetch_worker(
+      ss::shard_id shard,
+      op_context& octx,
+      shard_fetch fetch,
+      size_t min_fetch_bytes) {
+        // if over budget skip the fetch.
+        if (octx.bytes_left <= 0) {
+            co_return;
+        }
+        // no requests for this shard, do nothing
+        if (fetch.empty()) {
+            co_return;
+        }
+
+        const bool foreign_read = shard != ss::this_shard_id();
+
+        auto results = co_await octx.rctx.partition_manager().invoke_on(
+          shard,
+          octx.ssg,
```

Comment:
Can you comment/consider a bit about this SSG?

Previously, I think we just did "workfull" requests under the ssg, i.e., requests that just do work and don't wait. Now, we will be waiting while holding the ssg units, so they may become exhausted much more easily, blocking other fetchers or users of the ssg. We had something similar with the join/sync group calls in the group API as blocking during join group consumed the (5000-unit) ssg.

Reply:
This has been addressed in a comment below.